### PR TITLE
Add flash for logging in with banned users

### LINKED ISSUE #4691: Refactor alerts for Login flow
**We are preparing to participate in Google Code-in, and have reserved this issue for participants in GCI - but we'd love to have your help with another one! Please check out https://code.publiclab.org to see more.**

At time of login via OAuth the notices are wrongly shown and sometimes delayed too. See this for reference - https://drive.google.com/open?id=1w9tc4kIkoLJ5tvceoyQ1-kO_yug2pMDR

The link to file which requires help with notice - https://github.com/publiclab/plots2/blob/master/app/controllers/user_sessions_controller.rb#L58 

Also, this issue - #4522 can be referenced where details/progress about resolving the Login issue is recorded.

Thanks!

### PR DESCRIPTION
**Pull request by Uzay-G**
Partially fixes #4691

Logging in as a banned user with oauth was flashing "User was logged in." although he has been banned and does not log in. I fixed this by adding a conditional statement to check whether or not the user is banned or moderated. If he is, it flashes the corresponding message instead.

I tested this and everything works.
What do you think?

### CODE REVIEW

**Comment by SidharthBansal** [comment]
Awesome thanks

**Comment by SidharthBansal** [comment]
Kindly check Travis.

**Comment by SidharthBansal** [comment]
Kindly attach screenshot of the working and minitests

**Comment by Uzay-G** [comment]
Thanks! I am trying to make a test but I don't understand how I can login through oauth in the tests. Could you help me?

**Comment by Uzay-G** [comment]
Also do you know why there are these issues with codeclimate? It's about changes I didn't make.

**Comment by Uzay-G** [comment]
Hey, when we merge them, can you merge this one first and then #7074 because they have some conflicts. Thanks!

**Comment by Uzay-G** [comment]
I know everything works but I am having a problem with the test. This is what I have in `user_sessions_controller_test.rb`:
```ruby
  test "logging in with banned user" do
    request.env['omniauth.auth'] =  OmniAuth.config.mock_auth[:facebook2]
    post :create
    assert_equal "Successfully linked to your account!",  flash[:notice]
    post :destroy
   # User.find(??).ban
    post :create
    assert_equal flash[:error], I18n.t('user_sessions_controller.user_has_been_banned', username: User.find(1).username).html_safe
  end
```
I don't know how to get the current user and ban him after I have created him through omniauth.

**Comment by SidharthBansal** [comment]
Please see to the existing banned user test for non oauth providers. They might help you.

**Comment by Uzay-G** [comment]
Yeah no I have but it's different because with oauth, it's a mock signup and you're not signup up with a real user in the test. In the normal signup, you are actually signing up a user in the database and so you can get that user and ban it.

**Comment by SidharthBansal** [comment]
Can we do it like this?

**Comment by Uzay-G** [comment]
What do you mean?
It's impossible to do it with normal login because everything already works fine with that part. These changes were for oauth

**Comment by SidharthBansal** [comment]
Please wait. I am reviewing your other prs

**Comment by SidharthBansal** [comment]
There are some conflicts here kindly rectify them

**Comment by SidharthBansal** [comment]
This task will contain 5 points as it is complicated

**Comment by SidharthBansal** [comment]
Can we aim at 15 points in total? You have already achieved 6 points for #4691. Let's do some more work for 9 points. What do you say?
This PR will contain 5 points.

**Comment by Uzay-G** [comment]
Yeah no problem. What would you like me to add?
Thanks again for helping me out :smile:

**Comment by SidharthBansal** [comment]
I added one suggestions in one of your merged PRs.
Kindly check the issue if anything is remaining.

**Comment by Uzay-G** [comment]
I think I should add a test for logging in normally with a banned user. I looked at the issue and think these changes solve all the problems. I will see if there are any other improvements I can make.

**Comment by SidharthBansal** [comment]
There are many functionalities related to Oauth
1. User initially SIGNED UP with non oauth OR
2. User initially SIGNED UP with oauth
then signs off and logins:
1. User LOGINs with Oauth thereafter OR
2. User LOGINS with non oauth thereafter
This results in
1. User may have  account without email-password form OR
2. User may have  account without oauth providers.
3. User has both
If we add Oauth provider while user is logged in then it is termed as `linking` of the provider.
Does this makes sense?

**Comment by Uzay-G** [comment]
Oh ok I get it. Some of those are already in the [user_sessions_test](https://github.com/publiclab/plots2/blob/master/test/functional/user_sessions_controller_test.rb) but I will work on the other ones. Should I make another PR?

**Comment by SidharthBansal** [comment]
For what?

**Comment by SidharthBansal** [comment]
Sorry I can't understand

**Comment by Uzay-G** [comment]
The tests for logging in through the website with publiclab or signing up with oauth or through the site have already been implemented on the file.

**Comment by SidharthBansal** [comment]
> The tests for logging in through the website with public lab or signing up with OAuth or through the site have already been implemented on the file.

These are already implemented

**Comment by SidharthBansal** [comment]
> Should I make another PR?

Let's complete this pr first. I will assign you some related work then

**Comment by Uzay-G** [comment]
Ok

**Comment by Uzay-G** [comment]
@codecov rebase

**Comment by Uzay-G** [comment]
Hey, I have been looking around and searching for a way to get the current user through omniauth so I can ban but there really doesn't seem to be a solution...

**Comment by SidharthBansal** [comment]
Try creating an account with normal sign up and logging in with oath.

**Comment by Uzay-G** [comment]
Hey @SidharthBansal! It was difficult but I finally figured it out! I am so happy :smile:. I will add more tests now that I understand how everything works. I had also noticed that there were still some errors and the message "Logged in" would sometimes display instead of "Banned..." in certain edge cases so I added the check at the start of the social login.

**Comment by Uzay-G** [comment]
Could you accept my codecov task so I can claim this one? Thanks!

**Comment by Uzay-G** [comment]
I've solved the problems of the PR and have added a test. I have also found some other tests I could add on a similar but not feature. Should I do them in a new pr?

**Comment by SidharthBansal** [comment]
> Could you accept my codecov task so I can claim this one? Thanks

Sure

**Comment by SidharthBansal** [comment]
> Should I do them in a new pr?

New PR

**Comment by SidharthBansal** [comment]
> I had also noticed that there were still some errors and the message "Logged in" would sometimes display instead of "Banned..." in certain edge cases so I added the check at the start of the social login.

Kindly explain in detail. Step by step. Login/signup is highly complicated stuff here

File: test/functional/user_sessions_controller_test.rb [hunk 1]
```diff
@@ -266,4 +266,14 @@ class UserSessionsControllerTest < ActionController::TestCase
     post :destroy
     assert_equal "Successfully logged out.",  flash[:notice]
   end
+  
+  test "logging in with banned user through oauth should fail" do
```
Comment (SidharthBansal) [inline test/functional/user_sessions_controller_test.rb]: Well done. Are the tests related to this PR(banned)? If yes, then add it to this pr. If tests are for other omniauth/oauth sequences then add them in a new pr.

File: app/controllers/user_sessions_controller.rb [hunk 2]
```diff
@@ -26,8 +26,14 @@ def handle_social_login_flow(auth)
     unless params[:hash_params].to_s.empty?
       hash_params = URI.parse("#" + params[:hash_params]).to_s
     end
-
-    if signed_in?
+    @user = User.where(email: auth["info"]["email"]) ? User.find_by(email: auth["info"]["email"]) : User.find(@identity.user.id)
+    if @user&.status&.zero?
+      flash[:error] = I18n.t('user_sessions_controller.user_has_been_banned', username: @user.username).html_safe
+      redirect_to root_url
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: redirection is not handled here . Kindly handle it

File: app/controllers/user_sessions_controller.rb [hunk 3]
```diff
@@ -26,8 +26,14 @@ def handle_social_login_flow(auth)
...
+    elsif @user&.status == 5
+      flash[:error] = I18n.t('user_sessions_controller.user_has_been_moderated', username: @user.username).html_safe
+      redirect_to root_url
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: redirection is not handled here

Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: Kindly handle it

File: app/controllers/user_sessions_controller.rb [hunk 4]
```diff
@@ -26,8 +26,14 @@ def handle_social_login_flow(auth)
     unless params[:hash_params].to_s.empty?
       hash_params = URI.parse("#" + params[:hash_params]).to_s
     end
-
-    if signed_in?
+    @user = User.where(email: auth["info"]["email"]) ? User.find_by(email: auth["info"]["email"]) : User.find(@identity.user.id)
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: Please explain here.

File: app/controllers/user_sessions_controller.rb [hunk 5]
```diff
@@ -60,7 +66,14 @@ def handle_social_login_flow(auth)
           session[:openid_return_to] = nil
           redirect_to return_to + hash_params
         else
-          redirect_to return_to + hash_params, notice: I18n.t('user_sessions_controller.logged_in')
+          if session[:openid_return_to] # for openid login, redirects back to openid auth process
+            return_to = session[:openid_return_to]
+            session[:openid_return_to] = nil
+            redirect_to return_to + hash_params
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: Why two times `redirect_to return_to + hash_params`?

File: app/controllers/user_sessions_controller.rb (hunk 2 above)
Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: What do you mean?

**Comment by SidharthBansal** [comment]
Initially user is at /wiki/xys then clicks login

**Comment by SidharthBansal** [comment]
We need to preserve /wiki/xyz so that after login person gets to the URL
/wiki/xyz

File: test/functional/user_sessions_controller_test.rb (hunk 1 above)
Comment (Uzay-G) [reply test/functional/user_sessions_controller_test.rb]: Yeah they are

File: app/controllers/user_sessions_controller.rb (hunk 5 above)
Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Just an error

File: app/controllers/user_sessions_controller.rb (hunk 3 above)
Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Yeah I think I fixed it

File: app/controllers/user_sessions_controller.rb (hunk 2 above)
Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Yeah I think I fixed it

File: app/controllers/user_sessions_controller.rb (hunk 4 above)
Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: It gets the user data either through auth parameters or through omniauth identity and then if it found an existing user, it checks if the user was banned or moderated before signing in. If it didn't find one, that means it is a new user.

Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: I have to think here

Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: if User.where(email: auth["info"]["email"]) is not present
... this means user is trying to SIGN IN to public lab for the first time. Before account creation @user can't be banned NOR moderated.
Does this makes sense?

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: No I know this is how it works. If it's the first time the user signs in the `@user` variable is nil. Then, the `&`'s in `if @user&.status&.zero?` check whether or not the user is nil and so whether or not the status is nil. It's the same as `if @user and @user.status.zero?` but codeclimate told me to use the `&.` operator. So if `@user` is nil, that whole block is ignored and the function continues, otherwise it checks if it is banned or moderated. I didn't know about it either but learned [here](https://www.rubyguides.com/2018/07/ruby-operators/#The_Safe_Navigator_Operator).

**Comment by Uzay-G** [comment]
I don't think there are any other problems @SidharthBansal :+1:

File: app/controllers/user_sessions_controller.rb [hunk 6]
```diff
@@ -26,8 +26,14 @@ def handle_social_login_flow(auth)
     unless params[:hash_params].to_s.empty?
       hash_params = URI.parse("#" + params[:hash_params]).to_s
     end
-
-    if signed_in?
+    @user = User.where(email: auth["info"]["email"]) ? User.find_by(email: auth["info"]["email"]) : @identity.user
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: changes should go inside `if not signed in`.

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Why?

File: app/controllers/user_sessions_controller.rb (hunk 4 above)
Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: Great to see that you learnt about this. I want to tell you that this logic should be inside !signed_in loop

File: app/controllers/user_sessions_controller.rb [hunk 7]
```diff
@@ -26,8 +26,14 @@ def handle_social_login_flow(auth)
...
+    if @user&.status&.zero?
+      flash[:error] = I18n.t('user_sessions_controller.user_has_been_banned', username: @user.username).html_safe
+      redirect_to return_to + hash_params
+    elsif @user&.status == 5
+      flash[:error] = I18n.t('user_sessions_controller.user_has_been_moderated', username: @user.username).html_safe
+      redirect_to return_to + hash_params
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: We need a test to ensure this redirection too.

File: app/controllers/user_sessions_controller.rb [hunk 8]
```diff
@@ -26,8 +26,14 @@ def handle_social_login_flow(auth)
     unless params[:hash_params].to_s.empty?
       hash_params = URI.parse("#" + params[:hash_params]).to_s
     end
-
-    if signed_in?
+    @user = User.where(email: auth["info"]["email"]) ? User.find_by(email: auth["info"]["email"]) : @identity.user
+    if @user&.status&.zero?
+      flash[:error] = I18n.t('user_sessions_controller.user_has_been_banned', username: @user.username).html_safe
+      redirect_to return_to + hash_params
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: We need a test to ensure this redirection too.

File: app/controllers/user_sessions_controller.rb (hunk 6 above)
Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: Do you know about time complexity?
If person is already signed in then there is no need of your logic. 
It should be time optimized. Banned OR moderator user is a rare case. Think in terms of million qps, then this will be a huge overhead. What do you think?

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Like this?
```ruby
if signed_in?
   # code that is already there for that case
else
   # changes
```

Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: Yes

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Oh yeah ok I understand

Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: Did you understand my view point?
It is important from scalability views

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Does the last commit fix it?
Yeah I totally understand that it's quicker to do it that way

File: app/controllers/user_sessions_controller.rb (hunk 8 above)
Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Yeah but the problem here is that as mentioned in #6233, oauth is not redirecting correctly. I will try to understand why

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: This is the test I have for now but it won't pass because none of the redirection for oauth is working:
```ruby
  test "if banned user logs in through omniauth, should redirect properly" do
    request.env['omniauth.origin'] = "/notes/liked"
    request.env['omniauth.auth'] =  OmniAuth.config.mock_auth[:github1]
    # create account
    post :create
    # logout
    post :destroy
    User.find_by(name: "bansal_sidharth309").ban
    post :create
    assert_redirected_to "/notes/liked"
  end
```

**Comment by Uzay-G** [comment]
I am going to make a new pull request to add some more tests for the user_sessions controller!

File: app/controllers/user_sessions_controller.rb (hunk 8 above)
Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: We can handle oauth redirection in other pr. That PR should be merged before this pr

**Comment by Uzay-G** [comment]
Hey @SidharthBansal! I figured out how to fix the oauth redirection issue but it is another GCI task. No one has claimed it so can I make a pull request to fix it and then we can merge this one?
I am so happy! :smile:

**Comment by SidharthBansal** [comment]
Things are connected so you may have to do a bunch of task for points.

File: app/controllers/user_sessions_controller.rb [hunk 9]
```diff
@@ -50,7 +50,14 @@ def handle_social_login_flow(auth)
...
       end
     else # not signed in
-      if @identity&.user.present?
+      @user = User.where(email: auth["info"]["email"]) ? User.find_by(email: auth["info"]["email"]) : @identity.user
+      if @user&.status&.zero?
+        flash[:error] = I18n.t('user_sessions_controller.user_has_been_banned', username: @user.username).html_safe
+        redirect_to root_url
+      elsif @user&.status == 5
+        flash[:error] = I18n.t('user_sessions_controller.user_has_been_moderated', username: @user.username).html_safe
+        redirect_to root_url
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: waiting for redirect to be rectified first

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Alright I fixed that problem

Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: we will come back here after merging your other prs

File: test/functional/user_sessions_controller_test.rb [hunk 10]
```diff
@@ -266,4 +266,14 @@ class UserSessionsControllerTest < ActionController::TestCase
...
+    request.env['omniauth.auth'] =  OmniAuth.config.mock_auth[:github1]
+    post :create
+    post :destroy
+    # beginning of email of mock user is its name
+    User.find_by(name: "bansal_sidharth309").ban
+    post :create
+    assert_equal flash[:error], I18n.t('user_sessions_controller.user_has_been_banned', username: "bansal_sidharth309").html_safe
```
[the remark this answers is not in the document]
Comment (Uzay-G) [reply test/functional/user_sessions_controller_test.rb]: Yeah I added the change to test redirection but it will fail until we merge the oauth pull request

**Comment by SidharthBansal** [comment]
Please stay connected we will come back to this pr THanks

File: app/controllers/user_sessions_controller.rb [hunk 11]
```diff
@@ -50,7 +50,14 @@ def handle_social_login_flow(auth)
         redirect_to return_to + hash_params, notice: "Already linked to another account!"
       end
     else # not signed in
-      if @identity&.user.present?
+      @user = User.where(email: auth["info"]["email"]) ? User.find_by(email: auth["info"]["email"]) : @identity.user
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: User.find_by(email: auth["info"]["email"]) : @identity.user

i dont think we need this. explain why we needed this?

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: We want to check if the user is banned before he goes through the login process. There are two ways to find the user, through the `@identity` or the auth parameters. This assignment will either assign the user to `User.where(email: auth["info"]["email"])` if it exists or `@identity.user`. Once we have the user, we can check if he was banned.

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: We want to check if the user is banned before he goes through the login process. There are two ways to find the user, through the @identity or the auth parameters. This assignment will either assign the user to User.where(email: auth["info"]["email"]) if it exists or @identity.user. Once we have the user, we can check if he was banned.

Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: Try to understand the relationships.
A `user` has many `identity`(ies). Identity is a concept. Provider is stored inside identity. User has a field email. Provider = {Facebook, Google, Twitter, Github}. Also, if a user has an identity (or a provider) then identity has email too. So, while creating the user account we already saved the user email id from identity email id. Does this makes sense?
So, I don't think we need `@identity.user`. What do you suggest?
Maybe I am wrong. Or i am missing something important here. Think carefully. It is complex code.

**Comment by SidharthBansal** [comment]
@ananyaarun @IshaGupta18 kindly review this. This is important pr. I can't take risk here. I am a human too. I can err too. So, better to take multiple review here.

File: app/controllers/user_sessions_controller.rb (hunk 11 above)
Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Oh no I totally understand how that works. But I need to write it that way because the `@identity` **only exists** if the user has already linked his account to oauth. If for example, a user who already has a normal account through the website decides to link it to oauth, he does not have an identity and the only way to log him in is through the `oauth` parameters.

**Comment by SidharthBansal** [comment]
Yes you understand it correctly. So what's the issue now?

**Comment by Uzay-G** [comment]
Oh there is no issue. I was just explaining why we need to get the user that way and we cant just assign `@user` to `@identity.user`

**Comment by SidharthBansal** [comment]
Kindly resolve conflicts. Till now 14 points content is merged but you have not received points for that in the field of Oauth

**Comment by SidharthBansal** [comment]
I can't understand the use case of @identity.user.
Kindly give an example
Sorry for taking so long

**Comment by SidharthBansal** [comment]
I have already given 15 points in last oauth task. so no points remainings.

**Comment by Uzay-G** [comment]
No problem. So basically there are two ways we can get the users:
- if the user already has on Oauth account, we use `@identity.user`: https://github.com/publiclab/plots2/blob/82dd373fb39926d793fffdaf4b9b669135b134f2/app/controllers/user_sessions_controller.rb#L53-L55

- Otherwise, the only way to get the user is through the email sent by Omniauth like this: https://github.com/publiclab/plots2/blob/82dd373fb39926d793fffdaf4b9b669135b134f2/app/controllers/user_sessions_controller.rb#L91-L93.

**Comment by SidharthBansal** [comment]
@Uzay-G kindly read the amendment of the marks I did in the last approved issue. Sorry about it.

**Comment by SidharthBansal** [comment]
https://codein.withgoogle.com/dashboard/tasks/5997273587122176/
Please claim this task

**Comment by SidharthBansal** [comment]
Can you please point out to the place where I gave some points for setting up the MK, SWB and plots2 integrated login flow and their gifs?

**Comment by Uzay-G** [comment]
Yeah we discussed here: https://github.com/publiclab/plots2/issues/3367

**Comment by SidharthBansal** [comment]
Can you please indicate which comment? It is difficult to find

**Comment by SidharthBansal** [comment]
Here it is https://github.com/publiclab/plots2/issues/3367#issuecomment-571757321
10 points.

**Comment by Uzay-G** [comment]
No problem, you mentioned it here: https://github.com/publiclab/plots2/issues/3367#issuecomment-571757321

**Comment by Uzay-G** [comment]
Um, with the tests and the fixes I don't really think this was a medium task. It was quite difficult to track down the issues and then especially figuring out how to test the functionality was very complex. But it's your call :+1:

File: test/functional/user_sessions_controller_test.rb [hunk 12]
```diff
@@ -267,6 +267,18 @@ class UserSessionsControllerTest < ActionController::TestCase
     assert_equal "Successfully logged out.",  flash[:notice]
   end
   
+  test "logging in with banned user through oauth should fail and redirect correctly" do
```
Comment (SidharthBansal) [inline test/functional/user_sessions_controller_test.rb]: Write one test for moderator and one test for banned user. So, one test is missing.

**Comment by SidharthBansal** [comment]
Try to understand the place where it will break
  `@user = User.where(email: auth["info"]["email"]) ? User.find_by(email: auth["info"]["email"]) : @identity.user`
Suppose you are new to public lab and log in to public lab with Github for the first time in your life. Then the code will try to find out a user associated with the identity of GH before saving a user associated with this identity to the pl db. Does this makes sense?

**Comment by SidharthBansal** [comment]
I was just trying to give you hints. So that you can think. I have approved it on the dashboard already as I trust you will complete this issue gracefully.

**Comment by SidharthBansal** [comment]
Please read login documentation at https://github.com/publiclab/plots2/blob/master/doc/LOGIN_SYSTEMS.md once more. I think you misunderstood something in 
https://github.com/publiclab/plots2/pull/7072#issuecomment-573451073

**Comment by Uzay-G** [comment]
> Try to understand the place where it will break
> `@user = User.where(email: auth["info"]["email"]) ? User.find_by(email: auth["info"]["email"]) : @identity.user`
> Suppose you are new to public lab and log in to public lab with Github for the first time in your life. Then the code will try to find out a user associated with the identity of GH before saving a user associated with this identity to the pl db. Does this makes sense?

Yeah no this is a good point and that's why I check if there is a user beforehand:
https://github.com/publiclab/plots2/blob/d72137f2a52abbd3cd31c0af1ae2a67e43e6a835/app/controllers/user_sessions_controller.rb#L54-L57

The `&` operator checks if `@user` is not nil.

**Comment by Uzay-G** [comment]
> Please read login documentation at https://github.com/publiclab/plots2/blob/master/doc/LOGIN_SYSTEMS.md once more. I think you misunderstood something in
> [#7072 (comment)](https://github.com/publiclab/plots2/pull/7072#issuecomment-573451073)

What do you mean? I think it works this way but I will look again at the documentation. All the possible situations for logging in with oauth should be handled here.

File: app/controllers/user_sessions_controller.rb [hunk 13]
```diff
@@ -50,7 +50,14 @@ def handle_social_login_flow(auth)
         redirect_to return_to + hash_params, notice: "Already linked to another account!"
       end
     else # not signed in
-      if @identity&.user.present?
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: Your code should go inside `if @identity&.user.present?`
Does this makes sense?

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: No because the identity does not exist if the user has not already signed in through oauth. If for example, the user has created an account through PL website, logged out, and was then banned. The next day, they try to create a new account through oauth and enter the same email they used for their banned PL account. Without my changes, the user's credentials would go through here: 

https://github.com/publiclab/plots2/blob/d72137f2a52abbd3cd31c0af1ae2a67e43e6a835/app/controllers/user_sessions_controller.rb#L97-L103
The user would be linked to his **banned** PL account and then logged in although he is banned. That's why I can't put the code in the `@identity` section.

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Do you see what I mean?

Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: great!

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Thanks! I have really though about the signup and login logic a lot since I know how complex this issue is :+1:

**Comment by SidharthBansal** [comment]
Please add one more test
Add comment saying
User A has provider facebook with email xyz@abc.com while user has signed up for the first time with email xyzz@abcc.com is handled by @identity.user

**Comment by SidharthBansal** [comment]
Try to be more declarative. I write in a hurry

**Comment by Uzay-G** [comment]
> Please add one more test
> Add comment saying
> User A has provider facebook with email [xyz@abc.com](mailto:xyz@abc.com) while user has signed up for the first time with email [xyzz@abcc.com](mailto:xyzz@abcc.com) is handled by @identity.user

Sorry I don't understand what test do you want me to write?

**Comment by SidharthBansal** [comment]
We need a test for moderated user and a test for banned users

**Comment by SidharthBansal** [comment]
Test for moderated user is missing. Please do it asap

**Comment by Uzay-G** [comment]
Yeah I just added it

**Comment by Uzay-G** [comment]
Hey @SidharthBansal I also had created two other tests for the `user_sessions_controller_test` but they are not directly related to this one. Should I open a new pr?

File: app/controllers/user_sessions_controller.rb (hunk 11 above)
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: Add comment explaining this:
`User U has Provider P linked to U. U has email E1 while P has email E2. So, User table can't find E2 provided from auth hash, hence U is found by the user of identity having E2 as email`

Comment (SidharthBansal) [reply app/controllers/user_sessions_controller.rb]: You can place this comment exactly too

**Comment by SidharthBansal** [comment]
Let;s focus on this pr related work first. Open issues/prs having only 1 logic. This is getting complex. Modularity removes confusions among highly interlinked but highly versatile pieces of tech

File: app/controllers/user_sessions_controller.rb (hunk 11 above)
Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: I added comments, what do you think?

File: app/controllers/user_sessions_controller.rb [hunk 14]
```diff
@@ -50,7 +50,9 @@ def handle_social_login_flow(auth)
         redirect_to return_to + hash_params, notice: "Already linked to another account!"
       end
     else # not signed in
-      @user = User.where(email: auth["info"]["email"]) ? User.find_by(email: auth["info"]["email"]) : @identity.user
+      # if user has the same email for his oauth login and his publiclab website, we can get him using
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: We can't write informal text in comments. I have given you text. Kindly paste it directly. Thanks

**Comment by SidharthBansal** [comment]
@VladimirMikulic please give it a review. So that I can quickly merge this

**Comment by SidharthBansal** [comment]
@Uzay-G kindly claim https://codein.withgoogle.com/dashboard/task-instances/5266798652424192/

**Comment by SidharthBansal** [comment]
@VladimirMikulic kindly review it so that I can merge

**Comment by SidharthBansal** [comment]
@Uzay-G can you please check with the redirections in this pr?

**Comment by Uzay-G** [comment]
> @Uzay-G can you please check with the redirections in this pr?

I have included checks for the redirection in the pr. Do you want me to provide a gif?

**Comment by VladimirMikulic** [comment]
@SidharthBansal let's wait for Travis.

File: app/controllers/user_sessions_controller.rb [hunk 15]
```diff
@@ -50,7 +50,16 @@ def handle_social_login_flow(auth)
         redirect_to return_to + hash_params, notice: "Already linked to another account!"
       end
     else # not signed in
-      if @identity&.user.present?
+      # User U has Provider P linked to U. U has email E1 while P has email E2. So, User table can't find E2 provided
+      # from auth hash, hence U is found by the user of identity having E2 as email
+      @user = User.find_by(email: auth["info"]["email"]) || @identity.user
+      if @user&.status&.zero?
+        flash[:error] = I18n.t('user_sessions_controller.user_has_been_banned', username: @user.username).html_safe
+        redirect_to return_to + hash_params
```
Comment (SidharthBansal) [inline app/controllers/user_sessions_controller.rb]: Make sure line 58 and line 61 works for both oauth and non oauth users

Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: This part of the code only affects oauth users. I will provide a gif.

**Comment by SidharthBansal** [comment]
Gifs will be great

**Comment by SidharthBansal** [comment]
The piece of documentation you wrote have pronouns. Pronouns cause ambiguity in understanding the text. Due to readability issues I asked you to change your documentation. Documentation should be factual, short, concise, precise, traceable, unambiguous, readable, understandable, simple, highly detailed and gender-neutral. In case you can't understand any term then we can discuss further. 

Pronouns having ambiguity due to their gender bias nature and they are untraceable. This is natural language problem. Documentation is always written from the point of view of a naive developer.

**Comment by SidharthBansal** [comment]
Gender neutral language means using `they` instead of `he` or `she`. We should consider all the three genders equally at PL

**Comment by Uzay-G** [comment]
> 

> The piece of documentation you wrote have pronouns. Pronouns cause ambiguity in understanding the text. Due to readability issues I asked you to change your documentation. Documentation should be factual, short, concise, precise, traceable, unambiguous, readable, understandable, simple, highly detailed and gender-neutral. In case you can't understand any term then we can discuss further.
> 
> Pronouns having ambiguity due to their gender bias nature and they are untraceable. This is natural language problem. Documentation is always written from the point of view of a naive developer.

No I get it. I just hadn't thought about that. Sorry

File: app/controllers/user_sessions_controller.rb (hunk 15 above)
Comment (Uzay-G) [reply app/controllers/user_sessions_controller.rb]: Moderated redirection:
redirection-oauth-moderated

Banned redirection:
redirect-oauth-banned

**Comment by SidharthBansal** [comment]
No problem. I troubled a lot @HarshKhandeparkar @VladimirMikulic regarding docs. You were already troubled with me from 2 weeks. So, I wrote the comment myself and asked you to copy and paste.

**Comment by SidharthBansal** [comment]
@VladimirMikulic kindly review the gifs after travis passes.
I will merge after your approval. It's great to see collab + mentorship going across GCI by GCI students too :-)

**Comment by SidharthBansal** [comment]
@Uzay-G can you open up an issue and discuss about the additional tests you wanna add? That will be great

**Comment by VladimirMikulic** [comment]
@Uzay-G prepare for fixing Travis.

**Comment by Uzay-G** [comment]
Travis should succeed now

**Comment by Uzay-G** [comment]
Alright it worked. Do you think everything is good @VladimirMikulic?

**Comment by SidharthBansal** [comment]
@jywarren this went more complex than we initially thought.

**Comment by SidharthBansal** [comment]
Thanks all of you
